"""
Omni-Loop Engine v2.16 - Reflexive Multimodal AGI

Motor de ciclos reflexivos con skill_draft containerizado (Phoenix-integrated).
Cada consulta pasa por 3 iteraciones m√°ximo:
1. Draft inicial (skill_draft gRPC)
2. Reflexi√≥n sobre draft
3. Correcci√≥n auto-generada

FILOSOF√çA:
- Cada token es una decisi√≥n consciente
- El loop SIEMPRE termina (max 3 iteraciones hard-coded)
- El sistema se auto-corrige antes de responder
- Fallback a LFM2 si loop falla (resiliencia)

PHOENIX INTEGRATION:
- skill_draft via gRPC: 6s ‚Üí 0.5s por iteraci√≥n (-92%)
- skill_image via gRPC: 0MB host RAM (procesamiento externo)
- Contenedores ef√≠meros: aislamiento + auditor√≠a

Autor: SARAi Dev Team
Fecha: 29 octubre 2025
Versi√≥n: 2.16.0
"""

import os
import time
import hashlib
import logging
from typing import Optional, List, Dict, Any
from dataclasses import dataclass, asdict
from pathlib import Path

logger = logging.getLogger(__name__)


@dataclass
class LoopConfig:
    """Configuraci√≥n del Omni-Loop"""
    max_iterations: int = 3  # CR√çTICO: L√≠mite hard-coded (NO modificar)
    draft_model: str = "Qwen3-VL-4B-Instruct-iq4_nl"  # Modelo para draft
    enable_reflection: bool = True  # Habilitar auto-correcci√≥n
    confidence_threshold: float = 0.85  # Umbral para terminar loop early
    temperature: float = 0.7  # Temperatura para generaci√≥n
    max_tokens: int = 256  # Tokens m√°ximos por iteraci√≥n
    timeout_per_iteration: float = 10.0  # Timeout gRPC (segundos)
    
    # Phoenix Skills
    use_skill_draft: bool = True  # Usar skill_draft container (vs local)
    use_skill_image: bool = True  # Usar skill_image container (vs local)


@dataclass
class LoopIteration:
    """Resultado de una iteraci√≥n del loop"""
    iteration: int
    response: str
    confidence: float
    corrected: bool
    latency_ms: float
    tokens_generated: int = 0
    tokens_per_second: float = 0.0
    source: str = "skill_draft"  # "skill_draft" | "lfm2" | "local"


class OmniLoop:
    """
    Motor de loops reflexivos con skill_draft containerizado
    
    GARANT√çAS:
    - Loop SIEMPRE termina (max 3 iteraciones)
    - Fallback a LFM2 si falla (0% crash rate)
    - Latencia P50 <7.2s (target v2.16)
    - RAM P99 <9.6GB (skill_draft aislado)
    
    EJEMPLO:
    ```python
    loop = OmniLoop()
    result = loop.execute_loop(
        prompt="Explica la relatividad general",
        image_path="diagrams/spacetime.webp",
        enable_reflection=True
    )
    
    print(result["response"])  # Respuesta final auto-corregida
    print(result["auto_corrected"])  # True si hubo correcci√≥n
    print(result["iterations"])  # Lista de iteraciones
    ```
    """
    
    def __init__(self, config: LoopConfig = None):
        """
        Args:
            config: Configuraci√≥n del loop (usa defaults si None)
        """
        self.config = config or LoopConfig()
        self.loop_history: List[Dict] = []  # Historia de loops ejecutados
        
        # Validar configuraci√≥n cr√≠tica
        if self.config.max_iterations != 3:
            logger.warning(
                f"‚ö†Ô∏è max_iterations={self.config.max_iterations} != 3. "
                "Esto puede violar KPIs de latencia. Reverting to 3."
            )
            self.config.max_iterations = 3
        
        logger.info(f"[OmniLoop] Inicializado - "
                   f"Reflection: {self.config.enable_reflection}, "
                   f"Max iter: {self.config.max_iterations}")
    
    def execute_loop(
        self, 
        prompt: str, 
        image_path: Optional[str] = None,
        enable_reflection: Optional[bool] = None
    ) -> Dict[str, Any]:
        """
        Ejecuta loop reflexivo con skill_draft
        
        Args:
            prompt: Consulta del usuario (texto)
            image_path: Ruta a imagen para procesamiento multimodal (opcional)
            enable_reflection: Sobrescribe config.enable_reflection si se proporciona
        
        Returns:
            {
                "response": str,  # Respuesta final
                "iterations": List[LoopIteration],  # Lista de iteraciones
                "total_latency_ms": float,  # Latencia total del loop
                "auto_corrected": bool,  # True si hubo auto-correcci√≥n
                "fallback_used": bool,  # True si se us√≥ LFM2
                "metadata": {
                    "total_tokens": int,
                    "avg_tokens_per_second": float,
                    "confidence_final": float
                }
            }
        
        Raises:
            Exception: Solo si fallback tambi√©n falla (extremadamente raro)
        """
        start_time = time.perf_counter()
        
        # Usar override si se proporciona
        use_reflection = enable_reflection if enable_reflection is not None else self.config.enable_reflection
        
        iterations: List[LoopIteration] = []
        current_response = ""
        
        try:
            # PASO 0: Preprocesar imagen si existe (skill_image)
            processed_image_path = None
            if image_path:
                processed_image_path = self._preprocess_image(image_path)
            
            # ITERATION 1: Initial Draft
            logger.info(f"[OmniLoop] Iteration 1/3: Initial draft")
            iter1 = self._run_iteration(
                prompt=prompt,
                image_path=processed_image_path,
                iteration=1,
                previous_response=None
            )
            iterations.append(iter1)
            current_response = iter1.response
            
            if not use_reflection:
                # Sin reflexi√≥n, retornar draft directamente
                logger.info("[OmniLoop] Reflection disabled, returning draft")
                result = self._build_result(iterations, start_time, fallback=False)
                self.loop_history.append(result)  # CR√çTICO: Guardar en historia
                return result
            
            # ITERATIONS 2-3: Self-Reflection & Correction
            for i in range(2, self.config.max_iterations + 1):
                logger.info(f"[OmniLoop] Iteration {i}/3: Reflection & correction")
                
                reflection_prompt = self._build_reflection_prompt(
                    original_prompt=prompt,
                    draft_response=current_response,
                    iteration=i
                )
                
                iter_n = self._run_iteration(
                    prompt=reflection_prompt,
                    image_path=None,  # Solo texto en reflexi√≥n
                    iteration=i,
                    previous_response=current_response
                )
                iterations.append(iter_n)
                
                # Early exit si confidence es suficientemente alta
                if iter_n.confidence >= self.config.confidence_threshold:
                    logger.info(f"[OmniLoop] Early exit at iteration {i} "
                               f"(confidence: {iter_n.confidence:.2f})")
                    current_response = iter_n.response
                    break
                
                current_response = iter_n.response
            
            result = self._build_result(iterations, start_time, fallback=False)
            self.loop_history.append(result)
            return result
        
        except Exception as e:
            # FALLBACK: LFM2-1.2B (blindaje de continuidad)
            logger.error(f"‚ùå Omni-Loop failed: {e}. Falling back to LFM2.")
            
            try:
                fallback_response = self._fallback_lfm2(prompt)
                
                result = {
                    "response": fallback_response,
                    "iterations": iterations,
                    "total_latency_ms": (time.perf_counter() - start_time) * 1000,
                    "auto_corrected": False,
                    "fallback_used": True,
                    "fallback_reason": str(e),
                    "metadata": {
                        "total_tokens": 0,
                        "avg_tokens_per_second": 0.0,
                        "confidence_final": 0.0
                    }
                }
                
                self.loop_history.append(result)
                return result
            
            except Exception as fallback_error:
                # Si incluso LFM2 falla, lanzar excepci√≥n
                logger.critical(f"üî• CRITICAL: LFM2 fallback tambi√©n fall√≥: {fallback_error}")
                raise Exception(f"Omni-Loop AND fallback failed: {e}, {fallback_error}")
    
    def _run_iteration(
        self, 
        prompt: str, 
        image_path: Optional[str],
        iteration: int,
        previous_response: Optional[str]
    ) -> LoopIteration:
        """
        Ejecuta una iteraci√≥n del loop con skill_draft containerizado
        
        PHOENIX INTEGRATION:
        - skill_draft via gRPC: 6s ‚Üí 0.5s por iteraci√≥n (-92%)
        - Connection pooling reutiliza cliente gRPC
        - Timeout adaptativo seg√∫n max_tokens
        
        Args:
            prompt: Prompt para esta iteraci√≥n
            image_path: Imagen pre-procesada (si existe)
            iteration: N√∫mero de iteraci√≥n (1-3)
            previous_response: Respuesta de iteraci√≥n anterior
        
        Returns:
            LoopIteration con respuesta, confidence, latencia, etc.
        """
        start_time = time.perf_counter()
        
        # Construir prompt con contexto previo
        full_prompt = self._build_full_prompt(prompt, previous_response)
        
        if self.config.use_skill_draft:
            # ‚úÖ PHOENIX: skill_draft via gRPC
            try:
                response_data = self._call_skill_draft(full_prompt, image_path)
                source = "skill_draft"
            except Exception as e:
                logger.warning(f"‚ö†Ô∏è skill_draft failed: {e}. Fallback to local LFM2.")
                response_data = self._call_local_lfm2(full_prompt)
                source = "lfm2"
        else:
            # Modo legacy: LFM2 local (sin container)
            response_data = self._call_local_lfm2(full_prompt)
            source = "lfm2"
        
        latency_ms = (time.perf_counter() - start_time) * 1000
        
        # Calcular confidence y detectar correcci√≥n
        confidence = self._calculate_confidence(
            response=response_data["text"],
            prompt=prompt,
            iteration=iteration
        )
        
        corrected = (
            previous_response is not None and 
            response_data["text"] != previous_response
        )
        
        return LoopIteration(
            iteration=iteration,
            response=response_data["text"],
            confidence=confidence,
            corrected=corrected,
            latency_ms=latency_ms,
            tokens_generated=response_data.get("tokens", 0),
            tokens_per_second=response_data.get("tokens_per_second", 0.0),
            source=source
        )
    
    def _call_skill_draft(self, prompt: str, image_path: Optional[str]) -> Dict:
        """
        Llama a skill_draft container via gRPC
        
        Args:
            prompt: Texto del prompt
            image_path: Imagen pre-procesada (opcional)
        
        Returns:
            {"text": str, "tokens": int, "tokens_per_second": float}
        """
        from core.model_pool import get_model_pool
        
        try:
            # Importar protobuf (lazy import para evitar dependencias circulares)
            from skills import skills_pb2
            
            pool = get_model_pool()
            draft_client = pool.get_skill_client("draft")
            
            # Construir request gRPC
            request = skills_pb2.GenReq(
                prompt=prompt,
                max_tokens=self.config.max_tokens,
                temperature=self.config.temperature,
                stop=["</response>", "\n\n\n", "Human:", "Assistant:"]
            )
            
            # Si hay imagen, a√±adir al request (multimodal)
            if image_path:
                with open(image_path, "rb") as f:
                    request.image_bytes = f.read()
            
            # gRPC call con timeout
            response_pb = draft_client.Generate(
                request, 
                timeout=self.config.timeout_per_iteration
            )
            
            logger.info(
                f"‚úÖ skill_draft: {response_pb.tokens_per_second:.1f} tok/s, "
                f"RAM: {response_pb.ram_mb:.1f}MB"
            )
            
            return {
                "text": response_pb.text.strip(),
                "tokens": response_pb.tokens_generated,
                "tokens_per_second": response_pb.tokens_per_second
            }
        
        except Exception as e:
            logger.error(f"‚ùå skill_draft gRPC failed: {e}")
            raise
    
    def _call_local_lfm2(self, prompt: str) -> Dict:
        """
        Llama a LFM2-1.2B local (fallback)
        
        Args:
            prompt: Texto del prompt
        
        Returns:
            {"text": str, "tokens": int, "tokens_per_second": float}
        """
        from core.model_pool import get_model_pool
        
        pool = get_model_pool()
        lfm2 = pool.get("tiny")  # LFM2-1.2B
        
        # Generar con llama-cpp-python
        result = lfm2(
            prompt,
            max_tokens=self.config.max_tokens,
            temperature=self.config.temperature,
            stop=["</response>", "\n\n\n"]
        )
        
        text = result["choices"][0]["text"].strip()
        tokens = result["usage"]["completion_tokens"]
        
        # Calcular tokens/s (estimado)
        # NOTA: llama-cpp-python no expone timing, estimamos ~20 tok/s en CPU
        tokens_per_second = 20.0  # Estimaci√≥n conservadora
        
        return {
            "text": text,
            "tokens": tokens,
            "tokens_per_second": tokens_per_second
        }
    
    def _fallback_lfm2(self, prompt: str) -> str:
        """
        Fallback de emergencia a LFM2-1.2B
        
        Args:
            prompt: Prompt original del usuario
        
        Returns:
            Respuesta generada por LFM2
        """
        try:
            result = self._call_local_lfm2(prompt)
            return result["text"]
        except Exception as e:
            # Si incluso LFM2 falla, retornar mensaje de error
            logger.critical(f"üî• CRITICAL: LFM2 fallback failed: {e}")
            return "Lo siento, no puedo procesar tu consulta en este momento. Por favor, intenta de nuevo."
    
    def _preprocess_image(self, image_path: str) -> str:
        """
        Preprocesa imagen con skill_image (Phoenix)
        
        PHOENIX INTEGRATION:
        - skill_image via gRPC: 0MB host RAM
        - Conversi√≥n a WebP
        - Perceptual hashing para cache
        
        Args:
            image_path: Ruta a imagen original
        
        Returns:
            Ruta a imagen pre-procesada (WebP)
        """
        if not self.config.use_skill_image:
            # Modo legacy: sin preprocesamiento
            return image_path
        
        try:
            from core.model_pool import get_model_pool
            from skills import skills_pb2
            
            pool = get_model_pool()
            image_client = pool.get_skill_client("image")
            
            # Leer imagen
            with open(image_path, "rb") as f:
                image_bytes = f.read()
            
            # gRPC call para preprocesamiento
            request = skills_pb2.ImageReq(
                image_bytes=image_bytes,
                target_format="webp",
                max_size=1024  # Max width/height
            )
            
            response_pb = image_client.Preprocess(request, timeout=5.0)
            
            # Guardar imagen procesada
            output_path = f"state/images/{response_pb.image_hash}.webp"
            os.makedirs("state/images", exist_ok=True)
            
            with open(output_path, "wb") as f:
                f.write(response_pb.image_bytes)
            
            logger.info(f"‚úÖ skill_image: {output_path}, hash: {response_pb.image_hash[:8]}")
            
            return output_path
        
        except Exception as e:
            logger.warning(f"‚ö†Ô∏è skill_image failed: {e}. Using original image.")
            return image_path
    
    def _build_full_prompt(self, prompt: str, previous_response: Optional[str]) -> str:
        """
        Construye prompt completo con contexto previo
        
        Args:
            prompt: Prompt base
            previous_response: Respuesta de iteraci√≥n anterior
        
        Returns:
            Prompt completo para esta iteraci√≥n
        """
        if previous_response is None:
            # Primera iteraci√≥n: prompt directo
            return prompt
        
        # Iteraciones 2-3: prompt de reflexi√≥n
        return f"""[Previous attempt]
{previous_response}

[Task]
Reflect on the previous response and improve it if needed.
If it's already good, you can keep it the same.

[Original question]
{prompt}

[Improved response]"""
    
    def _build_reflection_prompt(
        self, 
        original_prompt: str, 
        draft_response: str,
        iteration: int
    ) -> str:
        """
        Construye prompt de reflexi√≥n para iteraciones 2-3
        
        Args:
            original_prompt: Pregunta original del usuario
            draft_response: Respuesta draft de iteraci√≥n anterior
            iteration: N√∫mero de iteraci√≥n (2 o 3)
        
        Returns:
            Prompt de reflexi√≥n
        """
        if iteration == 2:
            # Primera reflexi√≥n: revisar coherencia y completitud
            return f"""[Draft response]
{draft_response}

[Task]
Review the draft response above for the question: "{original_prompt}"

Is it coherent, complete, and accurate? If not, provide an improved version.
If it's already good, you can keep it.

[Improved response]"""
        
        else:  # iteration == 3
            # Segunda reflexi√≥n: pulir estilo y tono
            return f"""[Draft response]
{draft_response}

[Task]
Final polish for the question: "{original_prompt}"

Check for:
- Clarity and conciseness
- Appropriate tone
- No redundancy

[Final response]"""
    
    def _calculate_confidence(
        self, 
        response: str, 
        prompt: str,
        iteration: int
    ) -> float:
        """
        Calcula confidence score de la respuesta
        
        Heur√≠sticas:
        - Longitud razonable (50-500 chars) ‚Üí +0.2
        - Sin repeticiones excesivas ‚Üí +0.2
        - Responde a la pregunta (keywords match) ‚Üí +0.3
        - Iteraci√≥n 3 (√∫ltima oportunidad) ‚Üí +0.3 bonus
        
        Args:
            response: Texto de respuesta generada
            prompt: Prompt original
            iteration: N√∫mero de iteraci√≥n (1-3)
        
        Returns:
            Score entre 0.0 y 1.0
        """
        score = 0.0
        
        # 1. Longitud razonable
        length = len(response)
        if 50 <= length <= 500:
            score += 0.2
        elif 500 < length <= 1000:
            score += 0.1  # Penalizar respuestas muy largas
        
        # 2. Sin repeticiones excesivas
        words = response.lower().split()
        if len(words) > 0:
            unique_ratio = len(set(words)) / len(words)
            if unique_ratio > 0.7:
                score += 0.2
        
        # 3. Keywords match (simplificado)
        prompt_words = set(prompt.lower().split())
        response_words = set(response.lower().split())
        overlap = len(prompt_words & response_words)
        if overlap >= min(3, len(prompt_words)):
            score += 0.3
        
        # 4. Bonus por √∫ltima iteraci√≥n
        if iteration == 3:
            score += 0.3
        
        return min(score, 1.0)  # Clamp a [0, 1]
    
    def _build_result(
        self, 
        iterations: List[LoopIteration], 
        start_time: float,
        fallback: bool
    ) -> Dict[str, Any]:
        """
        Construye resultado final del loop
        
        Args:
            iterations: Lista de iteraciones ejecutadas
            start_time: Timestamp de inicio del loop
            fallback: True si se us√≥ LFM2 fallback
        
        Returns:
            Diccionario con resultado completo
        """
        total_latency_ms = (time.perf_counter() - start_time) * 1000
        
        # Respuesta final es la √∫ltima iteraci√≥n
        final_response = iterations[-1].response if iterations else ""
        
        # Auto-correcci√≥n si hubo >1 iteraci√≥n y cambi√≥ la respuesta
        auto_corrected = (
            len(iterations) > 1 and 
            any(iter.corrected for iter in iterations)
        )
        
        # Metadata agregada
        total_tokens = sum(iter.tokens_generated for iter in iterations)
        avg_tps = (
            sum(iter.tokens_per_second for iter in iterations) / len(iterations)
            if iterations else 0.0
        )
        confidence_final = iterations[-1].confidence if iterations else 0.0
        
        return {
            "response": final_response,
            "iterations": [asdict(iter) for iter in iterations],
            "total_latency_ms": total_latency_ms,
            "auto_corrected": auto_corrected,
            "fallback_used": fallback,
            "metadata": {
                "total_tokens": total_tokens,
                "avg_tokens_per_second": avg_tps,
                "confidence_final": confidence_final,
                "num_iterations": len(iterations)
            }
        }
    
    def get_loop_history(self) -> List[Dict]:
        """
        Retorna historia de loops ejecutados
        
        Returns:
            Lista de resultados de loops (√∫ltimos N)
        """
        return self.loop_history[-100:]  # √öltimos 100 loops
    
    def clear_history(self):
        """Limpia historia de loops"""
        self.loop_history.clear()
        logger.info("[OmniLoop] Historia limpiada")


def get_omni_loop(config: LoopConfig = None) -> OmniLoop:
    """
    Factory function para obtener instancia singleton de OmniLoop
    
    Args:
        config: Configuraci√≥n opcional (usa defaults si None)
    
    Returns:
        Instancia de OmniLoop
    """
    global _omni_loop_instance
    
    if "_omni_loop_instance" not in globals():
        _omni_loop_instance = OmniLoop(config)
    
    return _omni_loop_instance
